### ISSUE TITLE
List element reference mixed

### ISSUE THREAD

**Issue by VDerGoW** [KybernetikGames/animancer#204]
# Environment

- Animancer Version Number: 7.2
- Animancer Pro or Lite: Lite
- Unity Version: 2020.3.0f1
- Platform: Windows 11 64bit

# Description

In List, modify a SerializeReference ITransition filed will disturb other list elements.

# Reproduction

Steps to reproduce the bug:

1. Use following codes, create an asset.
2. Add an element to list, and set the element to ClipTransition, and assign an animation clip.
3. Add an element to list by click "+" button(Unity will copy last element automatically).
4. Assign element 1 another animation clip.
5. You can see that element 0 changed when modify element 1.

**Code**:
```csharp
[UnityEngine.CreateAssetMenu(fileName = "AnimancerRepro", menuName = "Test/AnimancerRepro")]
public class AnimancerRepro : UnityEngine.ScriptableObject
{
    public System.Collections.Generic.List<Wrapper> transitions = new System.Collections.Generic.List<Wrapper>();
}

[System.Serializable]
public class Wrapper
{
    [UnityEngine.SerializeReference]
    public Animancer.ITransition transition;
}
```

![demo](https://user-images.githubusercontent.com/11539008/164882480-96785d99-1650-482a-a8d2-a27032f664a1.gif)

**Comment by VDerGoW** [comment]
In this situation, I also found another issue: When clear list elements(click Reset or set list size to 0), will got an error `Could not reach source property name 'transitions.Array.data[0]' while extracting diffs, the reference does not exist in the source serialized data`.
![clear_list_error](https://user-images.githubusercontent.com/11539008/164885583-f177654f-a614-4bf4-96a3-1359fba54958.gif)

**Comment by KybernetikGames** [comment]
Both of those seem to be Unity bugs. The first can be replicated with the following script that doesn't use Animancer and the second is fixed in Unity 2021.2 (possibly before that, but that's the one I tested with after getting the error in an older version).

```
using System;
using UnityEngine;

public class NewBehaviourScript : MonoBehaviour
{
    public MyClass[] refs;

    private void Reset()
    {
        refs = new MyClass[] { new MyClass() { transition = new Transition() } };
    }
}

[Serializable]
public class MyClass
{
    [SerializeReference] public Transition transition;
}

[Serializable]
public class Transition
{
    public int value;
}
```
With a regular serialized list, assigning an item to two elements of the list would deserialize two copies of it, but with a serialized reference list it would only deserialize one and give the list two references to the same object. And when you increase the size of a list in the Inspector it just clones the last item so evidently it's not even bothering to clone it and just adding another reference to the same object.

With Animancer's Transitions, you can work around it by just using the dropdown to select the type of the transition (even if you set it to the already selected one) because that causes it to create a new object so it won't reference the same thing any more.

I'll see if I can come up with a way to detect when two reference fields are actually referencing the same transition so I can show something to point that out.
